### ISSUE TITLE
Code line number are not shown in generated pdf

### ISSUE THREAD

**Issue by yfain** [asciidoctor/asciidoctor-pdf#224]
The linenums is a great feature! This works fine in the asciidoctor generated HTML:

[source, javascript, linenums]

But asciidoctor-pdf seems to ignore linenums and no line numbers are shown in the generated PDF.

**Comment by mojavelinux** [comment]
You are correct, line numbers have not yet been implemented in the PDF output. Hopefully, it's just a matter of passing this information on to the syntax highlighter.

**Comment by mojavelinux** [comment]
I've done an initial analysis.

For Pygments, we're going to have to do line numbers ourselves because Pygments does line numbers in a very HTML-centric way.

CodeRay handles line numbers after the syntax highlighted result is generated, so again, we'll need to add them manually.

Thus, in both cases, we can't rely on the line numbering feature in the highlighter. We have to postprocess the result to add them. But this gives us the right level of control anyway, so it's the best solution in the end.

**Comment by mojavelinux** [comment]
I submitted a very crude implementation in PR #226. I'm likely going to rewrite it before merging...I just wanted to experiment and get it out there.

**Comment by mojavelinux** [comment]
Line numbers are now supported if you use the `rouge` syntax highlighter. We'll still need to implement this for the other syntax highlighters.

**Comment by anandchida** [comment]
This works for me with rouge. When used with tags within external source code file, it would be nice to pull in the line numbers from the source itself.
For example, within my source file, my line numbers could be

53  // tag::repositories[]
54 dependencies {
55     classpath 'org.asciidoctor:asciidoctor-gradle-plugin:1.5.2'
56  }
57  // end::repositories[]

when using linenums, it would be nice to see the same line numbers in the document as well.

**Comment by mojavelinux** [comment]
@anandchida Perhaps we can declare this using a value on the linenums attribute. Something like:

```
[source,groovy,linenums=implicit]
----
include::build.gradle[tag=repositories]
----
```

It might be best to track this as a separate issue request so it doesn't get lost.

**Comment by anandchida** [comment]
Done! https://github.com/asciidoctor/asciidoctor/issues/1541

**Comment by mojavelinux** [comment]
Thanks!

This may require a change to Asciidoctor PDF too, but let's address it first in core and see what needs to happen here once we have that in place.

**Comment by jfthuong** [comment]
Hi,

Any news on the support of line number while using `pygments`?

**Comment by krlawrence** [comment]
Hi All, first of all thanks so much for Asciidoctor - it's a great tool.

I am writing a book on Graph Databases and Apache Tinkerpop/Gremlin. Doing it all as open source as an aside. Anyway, yesterday for the first time I started adding line numbers to my source listings. The HTML output looks great but as reported above there are no line numbers in the PDF output. I am using Pygments. I really like the output that asciidoctor-pdf generates - feels a lot more "book like" than HTML but the lack of line numbers really stands out when compared to the HTML. Given this issue was opened in 2015 - is there any news on when the real fix may get committed? Many thanks in advance. FWIW there are several cases where the PDF output varies from the HTML output and it is always related to source code highlighting so I assume this is a problem area in general.

**Comment by mojavelinux** [comment]
I believe PR #226 is now ready to go. It also implements support for the start attribute to modify the starting line number. Please verify the behavior is correct. Then, I will merge.

**Comment by keryell** [comment]
Is there a version of `asciidoctor`+`asciidoctor-pdf`+some highlighter that is capable of displaying line numbers?

**Comment by mojavelinux** [comment]
As stated above:

> Line numbers are now supported if you use the rouge syntax highlighter.

**Comment by keryell** [comment]
Thanks. I was diving into https://github.com/asciidoctor/asciidoctor-pdf/blob/master/lib/asciidoctor/pdf/converter.rb#L1726 and it was not clear why it was not working for me.
I need to be sure I am really using the latest version of everything.
Is there still a requirement for Python 2 with Pygments?

**Comment by mojavelinux** [comment]
While Pygments is now compatible with Python 3, the version that's bundled with pygments.rb is not. To use Python 3, you have to tell it to use an external implementation, which is described here: https://asciidoctor.org/docs/user-manual/#using-a-custom-pygments-installation I haven't tried that approach to use Python 3 for Pygments, personally. And it's a large reason why I recommend Rouge over Pygments.

If you have further questions, please ask them in the gitter channel at https://gitter.im/asciidoctor/asciidoctor or a new issue. A closed issue is not the right place for a discussion.

**Comment by keryell** [comment]
OK, I got `rouge` working with line numbering with master branch, so I guess I can go on with this.
Thanks.

**Comment by mojavelinux** [comment]
That's excellent news! I'm glad to hear it. I think that puts you on the
best path for success.
